### DISCUSSION TITLE
How do we edit remote projects like VS Code?

### DISCUSSION THREAD

**Discussion by trusktr** [pulsar-edit/.github#146] [Pulsar Support]
In VS Code, we can open a project on a folder of a remote location connected via SSH. VS Code treats this remote project the same as a local project, except that the only difference is the files are on a remote machine. Can we do this with Pulsar?

**Comment by savetheclocktower** [comment]
I've never had to open an entire _project_ that was located on a remote server, but I use [remote-atom](https://web.pulsar-edit.dev/packages/remote-atom) all the time for editing of _single files_ located on a remote server. All it takes is installation of [one of a number of `rmate` implementations](https://github.com/aurora/rmate) dropped on the remote server somewhere in the user's `PATH`, at which point `rmate foo.txt` in the SSH session will open your local Pulsar.

I know it's nowhere near what you're describing, but it's a start.

**Reply by trusktr** [reply]
Yeah, I can't really use that workflow. I need full project features like intellisense, type checking, run tasks (f.e. from package.json), etc, basically the same as if I had opened the folder locally, just that it is remote.

Basically the abstraction to the file system should allow remote or local, and the rest of the editor should not care, should just work the same.

It does need some extra handling in the implementation though: it needs to forward ports (f.e. if you run a server on the remote, you want to open it locally), and VS Code installs a vscode server on the remote to do some handling (presumably to run tasks, etc, forwarding results to the UI, etc).

I think this is probably a number one feature Pulsar should aim to add. This can open many possibilities for Pulsar, including embedding in web, and the file system abstraction editing in a web-based file system, or connected to a remote, etc, while keep the experience the same no matter where the connection is to.

I heavily rely on this feature of VS Code currently. I wish I could switch to Pulsar! I really love (LOVE) the UX, design, and hackability, and Vim Mode Plus in Atom is extremely good (better than Vim or Neovim in my opinion).

**Reply by Daeraxa** [reply]
What about https://web.pulsar-edit.dev/packages/remote-ftp? Probably won't support all that stuff but it would at least do a project folder.

Otherwise its a good point, I think its something that Lapce supports out of the box as well - https://docs.lapce.dev/get-started/remote-development

**Reply by trusktr** [reply]
That's a partial solution. Not the full experience. Here's VS Code as a web app and we can edit any folder on our computer:

https://vscode.dev

The file system stuff is abstracted, so that it doesn't matter if the workspace is from local, remote, folder selected in a web browser, etc. In case of using the browser filesystem API, it can't run a terminal or tasks on the machine, so it is a little more limited. It is only an editor at that point, it simply disables some features that are unavailable.

I think this is the direction that Pulsar should plan for the future.

The only reason we need Electron is as a wrapper for some additional features, but we shouldn't rely on it. When wrapped in Electron, we can use native Node file system APIs. When wrapped in a browser, the file system abstraction adapts.

**Reply by Daeraxa** [reply]
Atom has had a web version discussed tons of times and it has been brought up more than a few times here but I don't think this is coming any time soon. I know it was a hard no from the Atom team whereas I don't think that is the case here but I wouldn't hold my breath for anything soon.

**Comment by Raf-sns** [comment]
Hi, i use the deprecated "FTP-remote-edit" but it's deprecated. 
Same you, i really want to see this feature on Pulsar
